src/pyhf/infer/__init__.py

```diff
@@ -153,18 +153,27 @@ def hypotest(
         tensorlib.astensor(CLs),
     )
 
-    _returns = [CLs]
+    is_q0 = kwargs.get('test_stat', 'qtilde') == 'q0'
+
+    _returns = [CLsb if is_q0 else CLs]
     if return_tail_probs:
-        _returns.append([CLsb, CLb])
+        if is_q0:
+            _returns.append([CLb])
+        else:
+            _returns.append([CLsb, CLb])
     if return_expected_set:
         CLs_exp = []
         for n_sigma in [2, 1, 0, -1, -2]:
 
             expected_bonly_teststat = b_only_distribution.expected_value(n_sigma)
 
-            CLs = sig_plus_bkg_distribution.pvalue(
-                expected_bonly_teststat
-            ) / b_only_distribution.pvalue(expected_bonly_teststat)
+            if is_q0:
+                # despite the name in this case this is the discovery p value
```

Comment:
We should probably make things more clear. If we're trying to read the code and it isn't clear what p-values are what that's going to be be problematic in the future.

Reply:
I opened #1247 to perhaps iterate on the internal names.

Reply:
Okay, I would have preferred to get this in 1 go, but as you already have this follow up PR ready to go then we can just focus on that right after this.